spec.bs

```diff
@@ -9278,28 +9362,117 @@ batcher=] |batcher| and a [=trusted scoring signals request=] |request|:
 </div>
 
 <div algorithm>
+To <dfn>extract URLs from creative info</dfn> given an [=ordered set=] |creativeInfoSet|:
+
+1. Let |result| be a new empty [=list=] of [=strings=].
+1. [=set/For each=] |creativeInfo| in |creativeInfoSet|:
+  1. [=list/Append=] [=URL serializer|Serialized=] |creativeInfo|'s [=creative info/ad
+    descriptor=]'s [=ad descriptor/url=] to |result|.
+1. Return |result|.
+
+</div>
+
+<div algorithm>
+To <dfn>extract creative scanning metadata</dfn> given an [=ordered set=] |creativeInfoSet|:
 
-To <dfn>build trusted scoring signals url</dfn> given a [=URL=] |signalsUrl|, a [=list=] of
-[=strings=] |renderURLs|, an [=ordered set=] of [=strings=] |adComponentRenderURLs|, an
-{{unsigned short}} |experimentGroupId|, and an [=origin=] |topLevelOrigin|:
+1. Let |result| be a new empty [=list=] of [=strings=].
+1. [=set/For each=] |creativeInfo| in |creativeInfoSet|:
+  1. [=list/Append=] |creativeInfo|'s [=creative info/creative scanning metadata=] to |result|.
+1. Return |result|.
+
+</div>
+
+<div algorithm>
+To <dfn>extract buyers</dfn> given an [=ordered set=] |creativeInfoSet|:
+
+1. Let |result| be a new empty [=list=] of [=strings=].
+1. [=set/For each=] |creativeInfo| in |creativeInfoSet|:
+  1. [=list/Append=] the [=serialization of an origin|serialization=] of |creativeInfo|'s
+    [=creative info/interest group owner=] to |result|.
+1. Return |result|.
+
+</div>
+
+<div algorithm>
+To <dfn>extract buyer and seller reporting ids</dfn> given an [=ordered set=] |creativeInfoSet|:
+
+1. Let |result| be a new empty [=list=] of [=strings=].
+1. [=set/For each=] |creativeInfo| in |creativeInfoSet|:
+  1. [=list/Append=] |creativeInfo|'s [=creative info/buyer or seller reporting id=] to |result|.
+1. Return |result|.
+
+</div>
 
-Note: When trusted scoring signals fetches are not batched, |renderURLs|'s [=list/size=] is 1.
+<div algorithm>
+To <dfn>add trusted signals param if needed</dfn> given a [=string=] |param|, a [=list=] of
+[=strings=] |input| and a [=list=] of [=strings=] |queryParamsList|:
+
+1. If |input| [=list/is not empty=]:

```

Comment:
if |input| is a list of empty strings like ["", "", ""], do we still want to do the following? Or it does not matter?

Reply:
Yes.  We want ",," as the result, so all the correlated lists line up.